compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -626,7 +626,10 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 assert!(expr_opt.is_none() || self.tcx.sess.has_errors());
             }
 
-            ctxt.may_break = true;
+            // If we encountered a `break`, then (no surprise) it may be possible to break from the
+            // loop... unless the value being returned from the loop diverges itself, e.g.
+            // `break return 5` or `break loop {}`.
+            ctxt.may_break |= !e_ty.is_never();
```

Comment:
Is `is_never()` correct? I would've expected something more general.

Reply:
Like `self.diverges` (was looking for it).

Reply:
My understanding is that the value of the break expression (i.e. the `x` in `break x`) will always have type `!` if it is known to diverge (in which case we'll never exit the loop). If we break somewhere inside `x`, then `may_break` will be set to `true` already by the subexpression. In there's a case you can think of that for which this might not hold, I can add another test?

Reply:
I guess one example might be something like `break { return; () }`?

Reply:
Another question would be something like

```rust
enum Void { }
fn foo() -> Void { panic!() }
break foo()
```

(To be honest, I'm not entirely sure what behavior I expect in *that* case.)

Reply:
@nikomatsakis: I've added those examples, and also an analogous test with `!`, which behave how I would expect (i.e. return `Void` does not work, but returning `!` does).

Reply:
```suggestion
            ctxt.may_break |= !self.diverges.is_always();
```
I think this is the change that @eddyb was proposing and I suspect it will fix the `break { return 0; () }` example.